File: benchmark/rclnodejs/topic/publisher-stress-test.js
// Copyright (c) 2018 Intel Corporation. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* eslint-disable camelcase */
import { program } from 'commander';
import rclnodejs from '../../../index.js';

program
  .option('-s, --size <size_kb>', 'The block size', '1')
  .option('-r, --run <n>', 'How many times to run', '1')
  .parse(process.argv);

const options = program.opts();
const size = parseInt(options.size) || 1;
const totalTimes = parseInt(options.run) || 1;

async function main() {
  try {
    await rclnodejs.init();

    const startTime = process.hrtime.bigint();
    const node = rclnodejs.createNode('stress_publisher_rclnodejs');
    const publisher = node.createPublisher(
      'std_msgs/msg/UInt8MultiArray',
      'stress_topic'
    );

    let sentTimes = 0;

    const message = {
      layout: {
        dim: [
          { label: 'height', size: 10, stride: 600 },
          { label: 'width', size: 20, stride: 60 },
          { label: 'channel', size: 3, stride: 4 },
        ],
        data_offset: 0,
      },
      data: new Uint8Array(1024 * size).map((_, index) => index & 0xff),
    };

    node
      .getLogger()
      .info(
        `The publisher will publish a UInt8MultiArray topic (contains a size of ${size}KB array) ${totalTimes} times.`
      );

    // Publish messages in a controlled manner
    const publishMessages = () => {
      return new Promise((resolve) => {
        const publishNext = () => {
          if (sentTimes < totalTimes) {
            publisher.publish(message);
            sentTimes++;
            setImmediate(publishNext);
          } else {
            resolve();
          }
        };
        publishNext();
      });
    };

    await publishMessages();

    const endTime = process.hrtime.bigint();
    const diffNanos = endTime - startTime;
    const diffMillis = Number(diffNanos) / 1000000;
    const seconds = Math.floor(diffMillis / 1000);
    const milliseconds = Math.round(diffMillis % 1000);

    node
      .getLogger()
      .info(
        `Benchmark took ${seconds} seconds and ${milliseconds} milliseconds.`
      );

    await rclnodejs.shutdown();
  } catch (error) {
    console.error('Error in publisher stress test:', error);
    process.exit(1);
  }
}

main();

File: benchmark/rclpy/topic/publisher-stress-test.py
#!/usr/bin/env python3
# Copyright (c) 2017 Intel Corporation. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import math
import time

import rclpy
from rclpy.node import Node
from std_msgs.msg import UInt8MultiArray, MultiArrayDimension, MultiArrayLayout

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-s", "--size", type=int, default=1, help="The block size[kb]")
    parser.add_argument("-r", "--run", type=int, default=1, help="How many times to run")
    args = parser.parse_args()

    rclpy.init()
    
    try:
        node = Node('stress_publisher_rclpy')
        
        # Create publisher
        publisher = node.create_publisher(UInt8MultiArray, 'stress_topic', 10)
        
        # Prepare message
        width_dim = MultiArrayDimension()
        width_dim.label = 'width'
        width_dim.size = 20
        width_dim.stride = 60

        height_dim = MultiArrayDimension()
        height_dim.label = 'height'
        height_dim.size = 10
        height_dim.stride = 600

        channel_dim = MultiArrayDimension()
        channel_dim.label = 'channel'
        channel_dim.size = 3
        channel_dim.stride = 4

        layout = MultiArrayLayout()
        layout.dim = [width_dim, height_dim, channel_dim]
        layout.data_offset = 0

        msg = UInt8MultiArray()
        msg.layout = layout
        msg.data = [x & 0xff for x in range(1024 * args.size)]
        
        node.get_logger().info(f'The publisher will publish a UInt8MultiArray topic '
                              f'(contains a size of {args.size}KB array) {args.run} times.')
        
        start_time = time.time()
        
        # Publish messages
        for i in range(args.run):
            publisher.publish(msg)
            rclpy.spin_once(node, timeout_sec=0.001)
        
        # Log completion
        diff = time.time() - start_time
        milliseconds, seconds = math.modf(diff)
        node.get_logger().info(f'Benchmark took {int(seconds)} seconds and '
                              f'{round(milliseconds * 1000)} milliseconds.')
        
        node.destroy_node()
    except KeyboardInterrupt:
        pass
    finally:
        rclpy.try_shutdown()

if __name__ == '__main__':
    main()
